packages/runtime-core/src/componentRenderUtils.ts

```diff
@@ -119,7 +119,11 @@ export function renderComponentRoot(
     // to have comments along side the root element which makes it a fragment
     let root = result
     let setRoot: ((root: VNode) => void) | undefined = undefined
-    if (__DEV__ && result.patchFlag & PatchFlags.DEV_ROOT_FRAGMENT) {
+    if (
+      __DEV__ &&
+      result.patchFlag & PatchFlags.DEV_ROOT_FRAGMENT &&
+      result.children
+    ) {
```

Comment:
According to the reproduction https://jsfiddle.net/skirtle/be5o7zv9/, if you try to look at the render function generated by it in the template explorer: https://vue-next-template-explorer.netlify.app/#%7B%22src%22%3A%22%3Cdiv%3E%5Cn%20%20%20%20%3Ctransition%3E%5Cn%20%20%20%20%20%20%3Cdiv%3E%3C%2Fdiv%3E%5Cn%20%20%20%20%3C%2Ftransition%3E%5Cn%20%20%3C%2Fdiv%3E%22%2C%22options%22%3A%7B%22mode%22%3A%22module%22%2C%22prefixIdentifiers%22%3Afalse%2C%22optimizeImports%22%3Afalse%2C%22hoistStatic%22%3Atrue%2C%22cacheHandlers%22%3Afalse%2C%22scopeId%22%3Anull%2C%22inline%22%3Afalse%2C%22ssrCssVars%22%3A%22%7B%20color%20%7D%22%2C%22bindingMetadata%22%3A%7B%22TestComponent%22%3A%22setup-const%22%2C%22setupRef%22%3A%22setup-ref%22%2C%22setupConst%22%3A%22setup-const%22%2C%22setupLet%22%3A%22setup-let%22%2C%22setupMaybeRef%22%3A%22setup-maybe-ref%22%2C%22setupProp%22%3A%22props%22%2C%22vMySetupDir%22%3A%22setup-const%22%7D%7D%7D

You will notice that the child node of the Transition component has been hoisted, this results in the value of `div.patchFlag` being `-1`, and `-1 & PatchFlags.DEV_ROOT_FRAGMENT` will always be `true`, which is wrong. So here we need to make sure in advance that the value of `div.patchFlag` is greater than 0.